extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -71,21 +79,33 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
         this.isAmpAdElement();
   }
 
+  /**
+   * For testing purposes, this value is retrieved by function call.
+   */
+  getVisibilityState_() {
+    return viewerForDoc(this.getAmpDoc()).getVisibilityState();
+  }
+
   /** @override */
   getAdUrl() {
     const startTime = Date.now();
     const global = this.win;
+    const adClientId = this.element.getAttribute('data-ad-client');
     const slotId = this.element.getAttribute('data-amp-slot-index');
     const slotIdNumber = Number(slotId);
     const correlator = getCorrelator(this.win, slotId);
     const screen = global.screen;
     const slotRect = this.getIntersectionElementLayoutBox();
-    const visibilityState = viewerForDoc(this.getAmpDoc()).getVisibilityState();
+    const visibilityState = this.getVisibilityState_();
     const adTestOn = this.element.getAttribute('data-adtest') ||
         isInManualExperiment(this.element);
     const format = `${slotRect.width}x${slotRect.height}`;
-    return googleAdUrl(this, ADSENSE_BASE_URL, startTime, slotIdNumber, [
-      {name: 'client', value: this.element.getAttribute('data-ad-client')},
+    const prevFmts = sharedState.getPrevFmts();
+
+    sharedState.addFormat(format);
```

Comment:
Add comment indicating this should be done AFTER building the prevFmts parameter.  Or you could modify sharedState to have single function:  getAndUpdatePreviousFormats(format) where it is given the slot's format...